mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmail/Action.php

```diff
@@ -0,0 +1,134 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Automation\Integrations\MailPoet\Actions\SendWelcomeEmail;
+
+use MailPoet\Automation\Engine\Workflows\ActionInterface;
+use MailPoet\Automation\Engine\Workflows\Step;
+use MailPoet\Automation\Engine\Workflows\Workflow;
+use MailPoet\Automation\Engine\Workflows\WorkflowRun;
+use MailPoet\Automation\Integrations\MailPoet\Subjects\SegmentSubject;
+use MailPoet\Automation\Integrations\MailPoet\Subjects\SubscriberSubject;
+use MailPoet\Entities\NewsletterEntity;
+use MailPoet\Entities\SegmentEntity;
+use MailPoet\Entities\SubscriberEntity;
+use MailPoet\InvalidStateException;
+use MailPoet\Newsletter\NewslettersRepository;
+use MailPoet\Newsletter\Scheduler\WelcomeScheduler;
+use MailPoet\Newsletter\Sending\ScheduledTasksRepository;
+use MailPoet\Segments\SubscribersFinder;
+
+class Action implements ActionInterface {
+  /** @var WelcomeScheduler */
+  private $welcomeScheduler;
+
+  /** @var NewslettersRepository */
+  private $newslettersRepository;
+
+  /** @var ScheduledTasksRepository */
+  private $scheduledTasksRepository;
+
+  /** @var SubscribersFinder */
+  private $subscribersFinder;
+
+  public function __construct(
+    WelcomeScheduler $welcomeScheduler,
+    NewslettersRepository $newslettersRepository,
+    ScheduledTasksRepository $scheduledTasksRepository,
+    SubscribersFinder $subscribersFinder
+  ) {
+    $this->welcomeScheduler = $welcomeScheduler;
+    $this->newslettersRepository = $newslettersRepository;
+    $this->scheduledTasksRepository = $scheduledTasksRepository;
+    $this->subscribersFinder = $subscribersFinder;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:send-welcome-email';
+  }
+
+  public function getName(): string {
+    return __('Send Welcome Email', 'mailpoet');
+  }
+
+  public function validate(Workflow $workflow, Step $step, array $subjects = []): ValidationResult {
+    $result = new ValidationResult();
+    if (!isset($step->getArgs()['welcomeEmailId'])) {
+      $result->addError('welcomeEmailIdRequired', 'Step arguments did not include a welcomeEmailId.');
+    } else {
+      $welcomeEmailId = (int)$step->getArgs()['welcomeEmailId'];
+      $newsletter = $this->newslettersRepository->findOneById($welcomeEmailId);
+      if ($newsletter === null) {
+        $result->addError('welcomeEmailNotFound', sprintf("Welcome Email with ID '%s' not found.", $welcomeEmailId));
+      } else {
+        $type = $newsletter->getType();
+        if ($type !== NewsletterEntity::TYPE_WELCOME) {
+          $result->addError('newsletterMustBeWelcomeType', sprintf("Newsletter must be a Welcome Email. Actual type for newsletter ID '%s' was '%s'.", $welcomeEmailId, $type));
+        } else {
+          $result->setNewsletter($newsletter);
+        }
+      }
+    }
```

Comment:
I think I still have mixed feelings about the complete validation (meaning constructing results with all errors) using the validation result classes. I'm not sure if we'll ever need to display all these errors. It will be more like "offer this or that action after some kind of trigger or not". And I guess we can make use of the new validator.

There are two or three parts to this:

1. Validating the arg types (the code above). I think this could be done using [the validator](https://github.com/mailpoet/mailpoet/pull/4027). The question is how exactly — should we implement a `getArgsSchema` method for each action? Or should the actions inject the validator themselves?
2. Required subjects (the code below). I'm not sure if this needs a specific method to return required subject types or if it should be more "free-form" (like it is now) — I guess having a specific method is too restrictive, some actions may say "I need a subscriber and an order or a segment" and other complex stuff. But it kind of seems that for now, we could be fine with a simple true/false logic. Or do you see any use for more complex validation results?
3. Anything else. So if an action wants to validate that... I don't know... some plugin is installed, it may be good to allow that. So some kind of a "free-form" validation method may make sense, maybe covering 2. and 3.

Reply:
1. I like the idea of using the validator for arguments, and my initial reaction is that I would prefer to have something like `getArgsSchema` be part of the interface. It seems like that would make it easier/more consistent to be able to figure out which actions are available for a given set of parameters. 

I'm thinking we should probably try to limit the scope of this PR for now and decide things like more complex validations stuff in a separate ticket. Does that sound appropriate to you?

2. The specific methods are almost certainly overkill. I think it was more that I wanted the certainty of the return types, as well as only having a single place that needs to know what the subject's `string` key is. For example, what if we have some validation method that says yes, `true`, all the required subjects are there, but then the `run` method looks for one of those subjects using `mailpoet:subscribers` instead of `mailpoet:subscriber` as the key?  But it's not like we truly need code to help out with situations like that, and the added complexity certainly isn't worth it, especially considering how much more restrictive it makes things.

3. I think it's a great idea to allow for free form validation that could check for just about anything. That may be something that we would want to handle in a separate ticket though?